File: Parallel Programming/problem2prl.cpp
/*
Emiliano Delgadillo Osorio
3 de Junio de 2025
Parallel programming and concurrent programming
Problem #2 Parallel

TEST 1: 
Enter a positive integer number: 82
Enter the number of threads: 4
Computing sum of primes (82) with: 4 threads
Created thread: 0, Range: [2, 21]
Created thread: 1, Range: [22, 41]
Created thread: 2, Range: [42, 61]
Created thread: 3, Range: [62, 82]
Sum of primes: 791

TEST 2:
Enter a positive integer number: 10
Enter the number of threads: 4
Computing sum of primes (10) with: 4 threads
Created thread: 0, Range: [2, 3]
Created thread: 1, Range: [4, 5]
Created thread: 2, Range: [6, 7]
Created thread: 3, Range: [8, 10]
Sum of primes: 17

TEST 3: 
Enter a positive integer number: 115
Enter the number of threads: 4
Computing sum of primes (115) with: 4 threads
Created thread: 0, Range: [2, 29]
Created thread: 1, Range: [30, 57]
Created thread: 2, Range: [58, 85]
Created thread: 3, Range: [86, 115]
Sum of primes: 1593

*/

#include <iostream> 
#include <cmath>
#include <pthread.h>
#include <iomanip>

using namespace std;

// Structure for the data read from the command line
typedef struct inputs {
    int n;
    int threads;
} inputs_t;

// Structure for the data to be sent to the threads
typedef struct thread_data {
    int id;
    pthread_t tid;
    pthread_mutex_t * mutex;
    int start;
    int stop;
    long long * total_sum;
} thread_data_t;

bool is_prime(int x) {
    if (x < 2) {
        return false;
    }
    
    if (x == 2) {
        return true;
    }
    
    // For any number i from 2 to sqrt(x)
    int limit = (int)sqrt(x);
    for (int i = 2; i <= limit; i++) {
        if (x % i == 0) {
            return false;
        }
    }
    
    return true;
}

// Thread function that will compute the sum of primes in a specific range
void * compute_prime_sum_range(void * data) {
    // Cast the pointer to void into the type we actually use
    thread_data_t * local_data = (thread_data_t *)data;
    // This will be the local counter, independent in each thread
    long long local_sum = 0;
    
    for (unsigned long long i = local_data->start; i <= local_data->stop; i++) {
        if (is_prime(i)) {
            local_sum += i;
        }
    }
    // Lock access to the shared variable before modifying it.
    // This will be done only once per thread
    pthread_mutex_lock(local_data->mutex);
        // Dereference the pointer to the total variable
        (*local_data->total_sum) += local_sum;
    pthread_mutex_unlock(local_data->mutex);

    pthread_exit(NULL);
}

long long compute_prime_sum(inputs_t inputs) {
    long long range = (inputs.n - 1) / inputs.threads;  
    // Create an array of data structures
    thread_data_t * data = new thread_data_t[inputs.threads];
    // Create the mutex that will be shared by all threads
    pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
    // Main result variable that will be shared by all threads
    long long total_sum = 0;
    // Variable to store the return value of the pthread functions
    int status;

    // Start all the threads
    for (int i = 0; i < inputs.threads; i++) {
        // Fill the data for each of the threads
        data[i].id = i;
        data[i].total_sum = &total_sum;
        data[i].mutex = &mutex;
        
        // Compute the start and stop values used by each thread
        data[i].start = 2 + (range * i);
        if (i == inputs.threads - 1) {
            data[i].stop = inputs.n;
        } else {
            data[i].stop = 2 + (range * (i + 1)) - 1;
        }

        status = pthread_create(&data[i].tid, NULL, &compute_prime_sum_range, (void *)&data[i]);
        printf("Created thread: %d, Range: [%d, %d]\n", i, data[i].start, data[i].stop);
        if (status == -1) {
            perror("ERROR: pthread_create");
        }
    }


    // Wait for the threads to finish
    for (int i = 0; i < inputs.threads; i++) {
        status = pthread_join(data[i].tid, NULL);
        if (status == -1) {
            perror("ERROR: pthread_join");
        }
    }
    return total_sum;
}

// Verify if we got command line arguments
// Otherwise ask the user for them 
inputs_t get_args(int argc, char* argv[]) {
    inputs_t inputs;
    if (argc == 3) {
        inputs.n = atoi(argv[1]);
        inputs.threads = atoi(argv[2]);
    } else if (argc == 2) {
        inputs.n = atoi(argv[1]);
        inputs.threads = 4;  
    } else {
        cout << "Enter a positive integer number: ";
        cin >> inputs.n;
        cout << "Enter the number of threads: ";
        cin >> inputs.threads;
    }
    return inputs;
}

// Entry point for the program
int main(int argc, char* argv[]) {
    inputs_t inputs = get_args(argc, argv);
    
    cout << "Computing sum of primes (" << inputs.n << ") with: " << inputs.threads << " threads" << endl;
    
    long long prime_sum = compute_prime_sum(inputs);
    
    cout << "Sum of primes: " << prime_sum << endl;
    
    return 0;
}

File: Parallel Programming/problem2prl.exs
#Emiliano Delgadillo Osorio
#3 de Junio de 2025
#Parallel programming and concurrent programming
#Problem #2 Parallel

defmodule BasicTasks do
  @doc """
  Toy examples of how to use the Task module and its functions
  """
  # Function to be called by the threads
  def prime_checker(id) do
    IO.puts("Hello from thread id #{id}")
    if is_prime(id), do: id, else: 0
  end

  # Launch a task and wait for it to finish
  def test1() do
    IO.puts("MAIN THREAD START")
    Task.start(fn -> prime_checker(2) end)
    Task.start(fn -> prime_checker(3) end)
    Task.start(fn -> prime_checker(5) end)
    Task.start(fn -> prime_checker(7) end)
    IO.puts("MAIN THREAD FINISH")
  end

  # Launch a task and wait for it to finish
  def test2() do
    IO.puts("MAIN THREAD START")
    task1 = Task.async(fn -> prime_checker(2) end)
    task2 = Task.async(fn -> prime_checker(3) end)
    task3 = Task.async(fn -> prime_checker(5) end)
    task4 = Task.async(fn -> prime_checker(7) end)
    _res1 = Task.await(task1)
    |> IO.inspect()
    _res2 = Task.await(task2)
    |> IO.inspect()
    _res3 = Task.await(task3)
    |> IO.inspect()
    _res4 = Task.await(task4)
    |> IO.inspect()
    IO.puts("MAIN THREAD FINISH")
  end

  # Launch multiple threads
  def test3(n) do
    IO.puts("MAIN THREAD START")
    1..n
    |> Enum.map(&Task.async(fn -> prime_checker(&1) end))
    |> Enum.map(&Task.await(&1))
    |> IO.inspect()
    IO.puts("MAIN THREAD FINISH")
  end
end

defmodule SuperPrimeSum do
  @moduledoc """
  Computations to find the sum of all prime numbers up to n either in sequence or in parallel
  """

  @doc """
  Check if a number is prime
  """
  def is_prime(x) when x < 2, do: false
  def is_prime(2), do: true
  def is_prime(x) do
    limit = :math.sqrt(x) |> trunc()
    not Enum.any?(2..limit, fn i -> rem(x, i) == 0 end)
  end

  @doc """
  Simple version of a prime sum function that works sequentially
  """
  def prime_sum(n), do: do_prime_sum(2, n, 0)

  defp do_prime_sum(current, n, acc) when current > n, do: acc
  defp do_prime_sum(current, n, acc) do
    if is_prime(current) do
      do_prime_sum(current + 1, n, acc + current)
    else
      do_prime_sum(current + 1, n, acc)
    end
  end

  @doc """
  Function to sum all prime numbers between 'start' and 'stop'.
  This will be used to find a part of a whole prime sum.
  """
  def prime_sum_range({start, stop}), do: do_prime_sum_range(start, stop, 0)

  defp do_prime_sum_range(start, stop, acc) when start > stop, do: acc
  defp do_prime_sum_range(start, stop, acc) do
    if is_prime(start) do
      do_prime_sum_range(start + 1, stop, acc + start)
    else
      do_prime_sum_range(start + 1, stop, acc)
    end
  end

  @doc """
  Compute the sum of primes of a number using multiple threads
  """
  def prime_sum_parallel(n, threads) do
    step = div(n - 1, threads)  # Start from 2, so subtract 1
    # Create the lists of starting and ending values for the ranges
    starts = make_range(2, step, threads)
    stops = make_range(2 + step - 1, step, threads)
    stops = List.replace_at(stops, -1, n)

    # Join the lists into a list of tuples
    Enum.zip(starts, stops)
    # |> IO.inspect()
    # Send each of the tuples to its own task to compute a partial sum
    |> Enum.map(&Task.async(fn -> prime_sum_range(&1) end))
    # Wait for the results. They will become a list
    |> Enum.map(&Task.await(&1, :infinity))
    # |> IO.inspect()
    # Sum all the elements of the list to get the final result
    |> Enum.reduce(0, &(&1 + &2))
  end

  @doc """
  Create a list with values that start at 'init', and increment by 'step'
  as many times as 'iter'
  """
  def make_range(init, step, iters), do: do_make_range(init, step, iters, [])

  defp do_make_range(_init, _step, 0, res), do: Enum.reverse(res)
  defp do_make_range(init, step, iter, res), do: do_make_range(init + step, step, iter - 1, [init | res])
end

defmodule Main do
  def main(args) do
    # Check how many arguments were sent
    case args do
      # One argument, compute the prime sum sequentially
      [n] ->
        SuperPrimeSum.prime_sum(String.to_integer(n))
        |> IO.inspect(label: "Sequential sum of primes: #{n}")
        [n, t] ->
          SuperPrimeSum.prime_sum_parallel(String.to_integer(n), String.to_integer(t))
          |> IO.inspect(label: "Parallel sum of primes: #{n}")
      _ ->
        IO.puts("Usage: \n $ elixir prime_sum.exs <n> <threads>")
    end
  end
end

# Call the main function with the arguments used in the command line
Main.main(System.argv())
